### PR TITLE
Switch to determining atom chiral types from pseudo 3D instead of bond dirs

### PR DESCRIPTION
**Pull request by greglandrum**
The current RDKit code to determine atomic stereochemistry based on bond wedging and 2D atom positions is a mess of trigonometry and special cases. This PR simplifies things by switching to a scheme which uses pseudo-3D:
Atoms at the end of dashed/wedged bonds are set above/below the plane and the stereo is determined using the usual chiral volume calculation. This is simpler (and likely more efficient, though that doesn't really matter here) and should be more maintainable going forward. I think it will also make it easier to identify situations where stereo is ambiguous.

A word on result changes:
I did need to tweak the number of expected failures in the InChI unit tests. All of the additional failures are due to horrible structures where the stereo as drawn is ambiguous. They are all like this:
image

I don't find that overly concerning. :-)

There's also a fix in here for a bug in the CDXML parser that @bp-kelley came up with. I'm going to also do a separate PR for that.

### CODE REVIEW

**Comment by greglandrum** [comment]
@bp-kelley do you have time to look at this one?

**Comment by bp-kelley** [comment]
I'll look at it by the weekend.

File: Code/GraphMol/catch_graphmol.cpp [hunk 1]
```diff
@@ -2630,6 +2630,7 @@ M  V30 END CTAB
 M  END
 )CTAB"_ctab;
   REQUIRE(m1);
+  m1->debugMol(std::cerr);
```
Comment (bp-kelley) [inline Code/GraphMol/catch_graphmol.cpp:2633]: Do we need debugging any more?

File: Code/GraphMol/testChirality.cpp [hunk 2]
```diff
@@ -2297,6 +2297,7 @@ void testGithub87() {
     std::string fName = rdbase + "/Code/GraphMol/test_data/github87.mol";
     RWMol *m = MolFileToMol(fName);
     TEST_ASSERT(m);
+    m->debugMol(std::cerr);
```
Comment (bp-kelley) [inline Code/GraphMol/testChirality.cpp:2300]: Remove debugging if not needed.

File: Code/GraphMol/Chirality.cpp [hunk 3]
```diff
@@ -769,6 +444,213 @@ const Atom *findHighestCIPNeighbor(const Atom *atom, const Atom *skipAtom) {
...
+      // std::cerr << bondVects[2] << std::endl;
+      // std::cerr << bondVects[3] << std::endl;
+      // std::cerr << "------------" << std::endl;
+      // std::cerr << crossp1 << " " << dotp1 << std::endl;
+      // std::cerr << crossp2 << " " << dotp2 << std::endl;
+      // std::cerr << " !!! " << vol << " " << vol2 << std::endl;
+      if (fabs(vol) < 1e-4) {
+        if (fabs(vol2) < 1e-4) {
+          BOOST_LOG(rdWarningLog)
+              << "Warning: ambiguous stereochemistry - no volume - at atom "
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp:631]: Perhaps '- no chiral volume -'

File: Code/GraphMol/Chirality.cpp [hunk 4]
```diff
@@ -769,6 +444,213 @@ const Atom *findHighestCIPNeighbor(const Atom *atom, const Atom *skipAtom) {
...
+    //       C
+    //       *
+    //       1
+    //   This is ST-1.2.10 in the IUPAC guidelines
+    //
+    //  Case 2: all bonds are wedged in the same direction
+
+    if (nNbrs == 3) {
+      bool conflict = false;
+      if (bondVects[order[1]].z * bondVects[order[0]].z < -1e-4 &&
```
Comment (bp-kelley) [inline Code/GraphMol/Chirality.cpp:571]: 1e-4 is used a lot, should this be a constant or at least split up between distance, angle and volume for clarity?

**Review by bp-kelley — CHANGES_REQUESTED** [review]

File: Code/GraphMol/catch_graphmol.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/catch_graphmol.cpp:2633]: nope!

**Comment by greglandrum** [comment]
@bp-kelley those changes are made

**Comment by greglandrum** [comment]
@bp-kelley ok for me to merge this?

**Review by bp-kelley — APPROVED** [review]

**Comment by bp-kelley** [comment]
You bet!

**Pull request merged**